src/catalua.cpp

```diff
@@ -77,47 +92,279 @@ bool lua_report_error(lua_State *L, int err, const char *path) {
+            return 1;
+        }
+        const auto iter = READ_MEMBERS.find( key );
+        if( iter == READ_MEMBERS.end() ) {
+            // No such member or function
+            lua_pushnil( L );
+            return 1;
+        }
+        lua_remove( L, -1 ); // remove key
+        // userdata is still there (now on -1, where it is expected by the getter)
+        return iter->second( L );
+    }
+    /**
+     * Wrapper for the Lua __newindex entry in the metatable of the userdata.
+     * It looks up the name of the requested member in @ref WRITE_MEMBERS and (if found),
+     * calls the function that the entry refers to (which acts as a setter).
+     */
+    static int newindex( lua_State * const L )
+    {
+        // -3 is the userdata, -2 is the key (name of the member), -1 is the value
+        const char * const key = lua_tostring( L, -2 );
+        if( key == nullptr ) {
+            luaL_error( L, "Invalid input to __newindex: key is not a string." );
+        }
+        const auto iter = WRITE_MEMBERS.find( key );
+        if( iter == WRITE_MEMBERS.end() ) {
+            luaL_error( L, "Unknown attribute" );
+        }
+        lua_remove( L, -2 ); // key, userdata is still there, but now on -2, and the value is on -1
+        return iter->second( L );
+    }
+    /**
+     * This loads the metatable (and adds the available functions) and pushes it on the stack.
+     */
+    static void get_metatable( lua_State* const L )
+    {
+        // Create table (if it does not already exist), pushes it on the stack.
+        // If the table already exists, we have already filled it, so we can return
+        // without doing it again.
+        if( luaL_newmetatable( L, METATABLE_NAME ) == 0 ) {
+            return;
+        }
+        // Push the metatable itself, the stack now contains two pointers to the same metatable
+        lua_pushvalue( L, -1 );
+        // Set the metatable of the new metatable (-2 on the stack) to be itself. Pretty meta, hu?
+        // It also pops one value from the stack.
+        lua_setmetatable( L, -2 );
+        // Now set the actual functions of the metatable.
+        luaL_setfuncs( L, &FUNCTIONS[0], 0 );
+
+        // Calling the destructor is really only needed when it's not trivial (e.g. pointers)
+        if( !std::is_trivially_destructible<T>::value ) {
+            // Push function pointer
+            lua_pushcfunction( L, &gc );
+            // -1 would be the function pointer, -2 is the metatable, the function pointer is popped
+            lua_setfield( L, -2, "__gc" );
+        }
+        lua_pushcfunction( L, &index );
+        lua_setfield( L, -2, "__index" );
+        lua_pushcfunction( L, &newindex );
+        lua_setfield( L, -2, "__newindex" );
+    }
 
-        luah_setglobal(L, "player", -1);
+public:
+    static void load_metatable( lua_State* const L )
+    {
+        get_metatable( L );
+        std::string mt(METATABLE_NAME);
+        if( mt.length() > 10 ) {
```

Comment:
Because THEY may change the rules at any time. THEY are known to do that. I don't take any risk with THEM, ever.

Also because string manipulation without explicit range checking always feels uneasy. It's like juggling with loaded guns, You assume the safety thing is on and it might work, but it may also fail horribly. This is start up code, it runs once, basically the opposite of time critical.